tests/test_upload.py

```diff
@@ -444,3 +447,73 @@ def test_upload_registration_fail(rse, scope, upload_client_registration_fail, f
             'did_name': fn,
             'guid': generate_uuid()
         }])
+
+
+@pytest.mark.parametrize("structure,failure", [
+    ({"container": ["file1", "file2", {"dataset": ["file3", "file4"]}]}, True),  # Cannot have both files and datasets in a container
+    ({
+        "container": {
+            "dataset": ["file1", "file2"]
+            },
+        "container2": {
+            "dataset": ["file1", {"dataset2": ["file1"]}]
+        }
+    }, True),  # One valid container, another invalid container
+    ({
+        "container": {"container2": {"dataset": {"dataset2": ["file1", {"dataset3": ["file2"]}]}}},
+    }, True),  # Same issue but further down the tree
+    ({"container": {"dataset": ["file1"], "dataset2": ["file2"]}}, False),  # One valid container
+    ({
+        "container": {"dataset": ["file1"], "dataset2": ["file2"]},
+        "container2": {"dataset": ["file1"], "dataset2": ["file2"]}
+    }, False)  # Two valid containers
+])
+def test_upload_recursive(structure, failure, rse_factory, scope, upload_client, caplog):
```

Comment:
I would change `failure` to something more descriptive like `should_fail`. I also think it might be better to make this a positive bool instead, something like `is_valid_structure`